README.md
# Lelylan Nodes

Enable a two way communication between the cloud and the physical world. Lelylan Nodes want to integrate the largest number of connected devices to a unique platform, making them all accessible through a unique API. A very nice work in this direction is [thethingsystem.com](http://thethingsystem.com/).

#### Supported Nodes

- [x] MQTT


#### Programmed Nodes

- [ ] Nest
- [ ] Lifx

Help us on adding [new nodes](http://thethingsystem.com/dev/supported-things.html).


## Requirements

Lelylan Nodes is tested against Node 0.8.8.


## Installation

    $ mongod
    $ git clone git@github.com:lelylan/mqtt.git
    $ npm install && npm install -g foreman
    $ nf start


## Resources

* [Lelylan Dev Center](http://dev.lelylan.com)
* [Mosca](https://github.com/mcollina/mosca)
* [Ascoltatori](https://github.com/mcollina/ascoltatori)


## Contributing

Fork the repo on github and send a pull requests with topic branches.
Do not forget to provide specs to your contribution.


### Running specs

* Fork and clone the repository
* Run `npm install`
* Run `npm test`


## Coding guidelines

Follow [Felix](http://nodeguide.com/style.html) guidelines.


## Feedback

Use the [issue tracker](http://github.com/lelylan/nodes/issues) for bugs or [stack overflow](http://stackoverflow.com/questions/tagged/lelylan) for questions.
[Mail](mailto:dev@lelylan.com) or [Tweet](http://twitter.com/lelylan) us for any idea that can improve the project.


## Links

* [GIT Repository](http://github.com/lelylan/nodes)
* [Lelylan Dev Center](http://dev.lelylan.com)
* [Lelylan Site](http://lelylan.com)


## Authors

[Andrea Reginato](https://www.linkedin.com/in/andreareginato)


## Contributors

Special thanks to all [contributors](https://github.com/lelylan/nodes/contributors)
for submitting patches.


## Changelog

See [CHANGELOG](https://github.com/lelylan/nodes/blob/master/CHANGELOG.md)


## License

Lelylan is licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).


CHANGELOG.md
# Changelog

## v0.1.0 (2016.01.01)

First node service version (only MQTT) for Lelylan.


test/helper.js
var Event        = require('../app/models/jobs/event')
  , User         = require('../app/models/people/user')
  , Application  = require('../app/models/people/application')
  , AccessToken  = require('../app/models/people/access_token');

// Remove all content on the used collections. Right now this is made only when the library
// is imported. Could be useful to create a function that can be called whenever you need it
// from the test suite to have a free of errors environment.
exports.cleanDB = function() {
  AccessToken.find().remove()
  User.find().remove()
  Application.find().remove()
};

// Set the event :realtime_processed field as true
exports.processed = function(event) {
  event.websocket_processed = true;
  event.save();
};


spec/app.spec.js
var request = require('supertest')
  , app = require('../app')
  , ascoltatori = require('ascoltatori')
  , async = require('async')
  , chai = require('chai')
  , expect = require('chai').expect
  , sinon = require('sinon')
  , debug = require('debug')('test');

chai.use(require('sinon-chai'));
chai.use(require('chai-fuzzy'));


var execute
  , ascoltatore
  , settings = {
      type: 'mongo',
      uri: process.env.MONGOLAB_JOBS_HOST,
      db: process.env.MONGOLAB_JOBS_DB,
      pubsubCollection: 'mqtt',
      mongo: {} };


describe('MQTT node',function() {

  describe('PUT /devices/:id/properties/set',function() {

    var payload = { id: 'device-1', properties: [{ id: 'property-1', value: 'on' }] };

    beforeEach(function(done) {
      ascoltatori.build(settings, function() { done() });
    });

    beforeEach(function() {
      execute = request(app)
        .put('/mqtt/devices/device-1/properties/set')
        .set('Content-Type', 'application/json')
        .set('X-Physical-Secret', 'secret-1')
        .send(payload)
    });

    it('returns 202', function(done) {
      execute.expect(202).expect({ status: 202 }, done);
    });

    it('publish on /mqtt/:secret/set', function(done) {
      async.series([
        function(cb) {
          ascoltatori.build(settings, function(ascoltatore) {
            ascoltatore.subscribe('mqtt/secret-1/set', function() {
              expect(arguments['0']).to.be.equal('mqtt/secret-1/set');
              expect(arguments['1']).to.be.like(payload)
              done()
            }); cb();
          });
        },
        function(cb) {
          execute.expect(202, cb)
        }
      ]);
    });
  });
});


test/app_spec.js
var request = require('supertest')
	, mongoose = require('mongoose')
  , app = require('../app')
  , ascoltatori = require('ascoltatori')
  , async = require('async')
  , chai = require('chai')
  , expect = require('chai').expect
  , sinon = require('sinon')
  , nock = require('nock')
  , debug = require('debug')('test');

chai.use(require('sinon-chai'));
chai.use(require('chai-fuzzy'));

var settings = {
  type: 'mongo',
  uri: process.env.MONGOLAB_JOBS_HOST,
  db: process.env.MONGOLAB_JOBS_DB,
  pubsubCollection: 'mqtt',
  mongo: {} };

var Factory = require('factory-lady')
  , Device = require('../app/models/devices/device');

require('./factories/devices/device');


describe('MQTT Node',function() {

  var execute, lelylan, device;
  var payload = { properties: [{ id: 'property-1', value: 'on' }] };

  beforeEach(function() {
    nock.cleanAll();
    Device.find().remove()
  });

  beforeEach(function(done) {
    Factory.create('device', {}, function(doc) {
      device = doc;
      done();
    });
  });


	describe('when receives a subscribed notification',function() {

		beforeEach(function() {
			lelylan = nock('http://api.lelylan.com')
				.matchHeader('accept', 'application/json')
				.matchHeader('X-Physical-Secret', device.secret)
				.filteringRequestBody(function(path) { return '*' })
				.put('/devices/' + device.id + '/properties', '*')
				.reply(200, {});
		});

		beforeEach(function(done) {
			ascoltatori.build(settings, function(ascoltatore) {
				ascoltatore.publish('devices/' + device.id + '/set', payload);
				done();
			});
		});

		it('sends a request to lelylan', function(done) {
			async.until(
				function() {
					return lelylan.isDone();
				},
				function(callback) {
					setTimeout(callback, 1);
				}, done);
		});
	});


	describe('when receives a PUT /mqtt/devices/:id',function() {

		beforeEach(function(done) {
			ascoltatori.build(settings, function() { done() });
		});

    describe('with valid x-device-secret header', function() {

      beforeEach(function() {
        execute = request(app)
          .put('/mqtt/devices/' + device.id)
          .set('Content-Type', 'application/json')
          .set('X-Physical-Secret', device.secret)
          .send(payload)
      });

      it('returns 202', function(done) {
        execute.expect(202).expect({ status: 202 }, done);
      });

      it('publish on /devices/:id/get', function(done) {
        async.series([
          function(cb) {
            ascoltatori.build(settings, function(ascoltatore) {
              ascoltatore.subscribe('devices/' + device.id + '/get', function() {
                expect(arguments['0']).to.be.equal('devices/' + device.id + '/get');
                expect(arguments['1']).to.be.like(payload)
                done()
              });
              cb();
            });
          },
          function(cb) {
            execute.expect(202, cb)
          }
        ]);
      });
    });

    describe('with no valid x-device-secret header', function() {

      beforeEach(function() {
        execute = request(app)
          .put('/mqtt/devices/' + device.id)
          .set('Content-Type', 'application/json')
          .set('X-Physical-Secret', 'not-valid')
          .send(payload)
      });

      it('returns 401', function(done) {
        execute.expect(401).expect({ status: 401 }, done);
      });

      it('does not publish on /devices/:id/get', function(done) {
        var spy = sinon.spy(ascoltatori, 'publish');
        async.series([
          function(cb) {
            execute.expect(401, cb)
          },
          function(cb) {
            expect(spy.callCount).to.have.eql(0);
            done();
          }
        ]);
      });
    });
	});
});


lib/loop.js
var mongoose = require('mongoose')
  , _ = require('underscore');

var Event        = require('../app/models/jobs/event')
  , User         = require('../app/models/people/user')
  , Application  = require('../app/models/people/application')
  , AccessToken  = require('../app/models/people/access_token');


if (process.env.DEBUG) { console.log('LELYLAN DEBUG: websocket worker up and running') }

// Find the valid tokens associated to property-udpated events
exports.execute = function(io) {
  Event.find({ websocket_processed: false, event: 'property-update' })
    .tailable().stream().on('data', function(collection) { findTokens(collection, io) });
}

// Returns all valid access tokens and notifies the apps using them
var findTokens = function(event, io) {

  // Find the subscriptions related to the resource owner active tokens
  var emit = function(err, tokens) {
    if (err) { console.log("LELYLAN ERROR", err.message) }
    if (process.env.DEBUG) { console.log('LELYLAN DEBUG: refreshing', tokens.length, 'dashboards') }

    _.each(tokens, function(token) { io.sockets.emit(token.token, { data: event.data, token: event.token }); });
    setWebsocketProcessed();
  }

  // Set the websocket_processed field to true
  var setWebsocketProcessed = function() {
    event.websocket_processed = true
    event.save()
  }

  // Find the access token that belongs to the user (valid clients)
  if (process.env.DEBUG) { console.log('LELYLAN DEBUG: processing event', event.resource_id) };
  event.findAccessTokens(emit)
}


app/models/jobs/event.js
var mongoose = require('mongoose')
  , db = mongoose.createConnection(process.env.MONGOLAB_JOBS_URL);

AccessToken = require('../people/access_token');

var eventSchema = new mongoose.Schema({
  resource_owner_id: mongoose.Schema.Types.ObjectId,
  resource_id: mongoose.Schema.Types.ObjectId,
  resource: String,
  event: String,
  token: String,
  source: String,
  data: mongoose.Schema.Types.Mixed,
  websocket_processed: { type: Boolean, default: false }
});


// Find the access token that belongs to the user (valid clients)
// See http://stackoverflow.com/questions/13279992/complex-mongodb-query-with-multiple-or/13280188
eventSchema.methods.findAccessTokens = function(callback) {

  AccessToken.find({
    resource_owner_id: this.resource_owner_id,
    revoked_at: undefined,
    $and: [
      { $or: [{ scopes: /resources/i }, { scopes: new RegExp(this.resource,'i') }] },
      { $or: [{ device_ids: { $size: 0 } }, { device_ids: this.resource_id }] }
    ]
  }, callback);
}


module.exports = db.model('event', eventSchema);


app.js
// nodetime monitoring
require('nodetime').profile({ appName: 'lelylan-nodes', accountKey: '030a1222bf8efceac3d583c62c31c5ae47ce1633' });
setInterval(function() { require('request')('http://nodes.lelylan.com'); }, 360000);

// libraries
var mongoose = require('mongoose')
  , express  = require('express')
  , app      = express()
  , server   = require('http').createServer(app)
  , request  = require('request')
  , debug    = require('debug')('lelylan')
  , Device   = require('./app/models/devices/device');

var mqtt = require('mqtt');

var host = process.env.MOSCA_HOST
  , port = '1883';

var ascoltatori = require('ascoltatori')
  , ascoltatore;

var settings = {
  type: 'redis',
  redis: require('redis'),
  db: 12,
  port: 6379,
  return_buffers: true,
  host: process.env.REDIS_HOST
};


// ---------------
// Express Server
// ---------------

app.configure(function() {
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(app.router);
  app.use(express.static(__dirname + '/app/assets'))
});

server.listen(process.env.PORT, function() {
  debug('Nodes server listening on port', process.env.PORT);
});


// ---------------
// Live page test
// ---------------

app.get('/', function(req, res) {
  res.sendfile(__dirname + '/index.html')
});


// -------------------------------------------
// From Lelylan to the Physical Device (/GET)
// -------------------------------------------

app.put('/mqtt/devices/:id', function(req, res) {
  var status = 401;
  debug('Receiving request', req.body);

  Device.findOne({ _id: req.params.id, secret: req.get('X-Physical-Secret') }, function (err, doc) {
    if (err) console.log(err.message);
    if (doc) {
      status = 202;
      publish(req, '/get/')

      // BAD HACK
      //var client = mqtt.createClient(port, host, { username: req.params.id, password: req.get('X-Physical-Secret') });
      //client.on('connect', function() {
        //client.publish('devices/' + req.params.id + '/get', JSON.stringify(req.body));
        //client.end();
      //})
    };

    res.status(status).json({status:status});
  });
});

var publish = function(req, mode) {
  // Needed to work with Redis (and Mosca default settings)
  // * Hex transformation for the JSON body
  // * Binary definition for Redis to understand the stored format type
  payload = { message: new Buffer(JSON.stringify(req.body)).toString('hex'), binary: true };
  var topic = 'devices/' + req.params.id + mode;
  ascoltatore.publish(topic, payload, function() {
    console.log('[API REQ] Message published to the topic', topic, payload);
  });
}


// -------------------------------------------
// From the Physical Device to Lelylan (/SET)
// -------------------------------------------

ascoltatori.build(settings, function (_ascoltatore) {
  ascoltatore = _ascoltatore;

  ascoltatore.subscribe('devices/*', function() {
    debug('TOPIC', arguments['0'], 'PAYLOAD', arguments['1']);
    var data = arguments['0'].split('/');
    if (data[2] == 'set')
      syncLelylan(data[1], arguments['1']);
  });
});

var syncLelylan = function(id, payload) {
  var uri = process.env.LELYLAN_API_URL + '/devices/' + id + '/properties';
  var json = JSON.parse(new Buffer(payload.message, 'hex').toString('utf8'));  // needed with Redis
  var options = { uri: uri, method: 'PUT', body: json, json: true };

	Device.findOne({ _id: id }, function (err, doc) {
		if (err) console.log('ERROR', err.message);

		if (doc) {
			options.headers = setHeaders(doc.secret);

			request(options, function(err, response, body) {
				if (err) console.log('ERROR', err.message);
				debug('Sent request to', uri, json)
			});
		}
  });
}

var setHeaders = function(secret) {
  return { 'X-Physical-Secret': secret, 'Content-Type': 'application/json' }
}

module.exports = app;
